### PR TITLE
Edit Medication Modal

### PR DESCRIPTION
**Pull request by Sectonic**
## Edit Medication Modal
Issue Number(s): #63 

### What does this PR change and why?
This PR mainly adds the new edit medication modal at the route `/edit-medication/[medicationId]`

### Checklist

- [x] The Edit Medication modal should exist at the following route: /edit-medication/{medicationId}
- [x] The flow should reuse the components utilized for the Add New Medication modal, with low coupling between modal and edit logic
- [x] The modal should first fetch the medication info, and display it in the "Review" section
- [x] If any of the pencils are clicked, it should THEN navigate to the respective page for editing
- [x] All changes, if the save button is clicked, should send a PATCH request to save the medication
- [x] Error handling should be implemented exhaustively
- [x] Test exhaustively for edge cases

### Critical Changes
- Created new base modal template that works for both add medication and edit medication
- Created the edit medication modal based on base modal template
- Fixed the modal data parser and implemented the model data parser 
- Changed the medication model to reflect better on the data passed in from the modal
- Fixed the zod implementation for medication

### Testing
For testing, I had MongoDBCompass open to see the ids of the medications I made at the route `/add-new-medication`, then with the ids I navigated to `/edit-medication/[id]` to test the modal.

### CODE REVIEW

**Comment by ShivaniVora** [comment]
This looks great so far! I just had some comments on some little things:
- For the Edit Medication flow (both in the Review stage and when someone has chosen to edit a field) the top says **Add New Medication** when it should say **Review** or **Edit Medication**
- When editing a field of the medication the user should be directed to a modal that is just one screen, rather than the relevant page in the Add New Medication flow (they shouldn’t have to click through all of the remaining editable fields to edit just one). This means the bottom should **not** say Next and lead to the next Add New Medication screen but rather say **Confirm** and **lead to the Review stage**. (The last two screens in the Figma say Add New Medication at the top and not Edit Medication, I believe this is an error and they should say Edit Medication but if that is not correct I will update you)
Screenshot 2025-03-13 at 1 27 48 PM


- When a user clicks Save without making any changes to the medication it should not throw an error rather it should just close the modal. We can add in a backend 204 No Content error. (I don't think this was mentioned earlier but should be an easy change hopefully)
- I think Cancel might be the wrong shade of red.
- We need to handle resizing, I would suggest looking at the breakpoints in tailwing_config.ts and try and use those to develop new sizes for the elements in the modal. You might need an additional one between Desktop and Tablet because of the width of the modal as well. (I didn't do this super cleanly but it is technically done in the login/signup/forgot password pages and if you need any help just reach out!)

Screenshot 2025-03-13 at 11 49 43 AM

File: src/components/modals/medication/medicationModalInfo.ts [hunk 1]
```diff
@@ -1,7 +1,14 @@
...
+export interface MedicationModalInfo {
+  /**
+   * If the Model ID is required, in Edit Modal
+   */
+  _id?: string;
+
   general: {
     /** Type of medication */
-    form: "Pill" | "Injection";
+    form: "tablet" | "liquid" | "injection";
```
Comment (michaelhyi) [inline src/components/modals/medication/medicationModalInfo.ts:11]:
```suggestion
    form: "tablet" | "liquid" | "injection";
```

In the actual frontend, it specifies pill, syrup, and shot. Can we make the datatypes consistent with those strings?

File: src/utils/parseAddMedition.ts
Comment (michaelhyi) [inline src/utils/parseAddMedition.ts:1]: Btw the file is misspelled.

File: src/utils/parseModalMedition.ts
Comment (michaelhyi) [inline src/utils/parseModalMedition.ts:1]: Btw this file is misspelled.

File: src/components/modals/medication/sections/dosageNotificationSection.tsx [hunk 2]
```diff
@@ -115,13 +115,13 @@ export default function DosageNotificationSection({
...
                 temp.dosage.notificationFrequency =
-                  newValue as AddMedicationInfo["dosage"]["notificationFrequency"];
+                  newValue as MedicationModalInfo["dosage"]["notificationFrequency"];
                 return temp;
               })
             }
           >
-            <Option value="Once / Day of Dosage" />
-            <Option value="Every Dose" />
+            <Option value="day of dose">Day of Dosage</Option>
```
Comment (michaelhyi) [inline src/components/modals/medication/sections/dosageNotificationSection.tsx:123]: Any reason for the migration from `Once / Day of Dosage => day of dose`? Just want to clarify all design discrepancies and if Cynthia + Michelle are aware.

File: src/components/modals/medication/medicationModalInfo.ts [hunk 3]
```diff
@@ -1,7 +1,14 @@
-export interface AddMedicationInfo {
+import { DayOfWeek } from "@/lib/consts";
+
+export interface MedicationModalInfo {
```
Comment (michaelhyi) [inline src/components/modals/medication/medicationModalInfo.ts:3]: Let's standardize this type more. This type on the frontend utilizes nested JSON objects, whereas the request body type on the backend uses a single flattened JSON object. Let's just use the same type on both sides, for simplicity.

**Review by michaelhyi — CHANGES_REQUESTED** [review]
Great work Sujal! Appreciate your hard work throughout out this sprint. Here are some comments I have:

- [ ] I noticed that you updated the review sections of both modals to be a singular scrollable view rather than the original one with two columns. It looks good, but deviates from the Figma designs, so it might not be what Cynthia and Michelle had in mind. Did you talk to them about this? And did you have any reasoning for this migration??
- [ ] It seems that creating a medication is still slightly buggy. Here's everything I caught: 
  - [ ] On submitting a medication that repeats every day or weekly, repeatMonthlyType defaults to week, repeatMonthlyOnWeekDay defaults to Sunday, and repeatMonthlyOnWeek defaults to 1. These values should not even exist, as it implies that its being repeated monthly
  - [ ] dosesUnit just says doses in the db. Do you mean day or hours?
  - [ ] There is no validation for checking for the same times being inserted for the same medication across two doses (ie 9:00AM and 9:00 AM again), and are we checking that doses should be in chronological order?
  - [ ] When I press take every X hours, the following error is thrown: `Failed to add medication: Medication validation failed: dosesPerDay: Path `dosesPerDay` is required.` dosesPerDay should only be required IF takeEvery blank hours is not selected.

File: src/components/modals/medication/sections/dosageNotificationSection.tsx (hunk 2 above)
Comment (Sectonic) [reply src/components/modals/medication/sections/dosageNotificationSection.tsx:123]: all of the changes to the values I made were from the one medication actions/controllers issue and all the sample bodies sent to those requests. I can change it back to the ones that are actually apart of the modal - I think that makes more sense anyways.

File: src/components/modals/medication/medicationModalInfo.ts (hunk 3 above)
Comment (Sectonic) [reply src/components/modals/medication/medicationModalInfo.ts:3]: Yeah I agree, ill start doing that

File: src/utils/parseModalMedition.ts
Comment (Sectonic) [reply src/utils/parseModalMedition.ts:1]: haha whoops

**Comment by ShivaniVora** [comment]
Hey Sujal! If any of these are not relevant due to changes that were agreed upon just ignore them! The modal looks great thank you for implementing the earlier comments, the only things I see now are I am missing the delete button on the review page as well as the back button on the edit pages, and I believe we can get rid of the dots at the bottom of the edit pages as they cannot switch between the pages anyways. Thank you so much it looks amazing!!

**Comment by cynthiazhang755** [comment]
Hey Sujal! Thanks so much for working on this ticket! It looks great! Just to echo Shivani's comments, 
1. Please double check that your modal matches the figma, I'm noticing that some of the edit medication pages do not all have a back button and a confirm button.
2. Please remove the progress dots on all pages. 
3. Please make sure that if the user clicks the back or the confirmation button, it reroutes the user back to the edit medication review page. 
4. Please add the cancel button to the bottom left-hand corner of the review page.
5. I also noticed that the add new medication modal got changed up (ie next button got changed to a confirm button), please make sure that the edit medication modal does not change the add new medication modal. 
I know this is a super quick turn around time for all these changes so no worries if you aren't able to finish everything by tonight's GM! We can def talk more about it during GM if you have any questions. Thanks so much again for your hard work!

**Comment by cynthiazhang755** [comment]
Hey Sujal thanks for fixing the navigation changes! There's a small loophole in the edit medication logic. Right now, if I change the "Take X doses per day" to let's say 5 and then click next to select the times and then click confirm, an error appears saying that no duplicate times are allowed. However, if I click back without changing any of the duplicate times, the modal saves the duplicate times and then when I click Save in the main review page, it causes a failure state. Could you change "The selecting dose times" page so that clicking the back button navigates to the previous "Take X doses per day" screen, and only the back button on the "Take X doses per day" screen can navigate back to the review page?
Screenshot 2025-03-26 at 4 37 54 PM

**Review by michaelhyi — CHANGES_REQUESTED** [review]
Great work Sujal! This works really well. Just some final, minor bugs I spotted:
- [ ] Although validating for duplicate times works on the frontend, it doesn't work on the backend!
Screenshot 2025-03-26 at 8 40 54 PM
- [ ] I get the following error every time I create a new medication. It goes away after clicking create twice though.
Screenshot 2025-03-26 at 8 49 59 PM
- [ ] The checkmark boxes are slightly uneven. See how the first one is slightly not a perfect box, while the second one is? Do you happen to know why?
Screenshot 2025-03-26 at 8 49 54 PM

**Comment by Sectonic** [comment]
I changed the naming of the buttons on edit because they weren't specific enough to what they did. For example one feature is that we want to go back to review on every single screen if we click back except the times screen which we want to go back to the dosage screen. In that instance we need something else than back to signify its different than the others. That's why I went with "Return to review" for other instances and in this case I made it "Back to Doses." And for saving the changes, I made the button name "Save Changes" instead of "Confirm" and for the dosage ->  times, I did "Proceed" instead of "Next".

**Review by michaelhyi — CHANGES_REQUESTED** [review]
Super close to merging Sujal! One tiny error:

image
When sending a request to update a medication, the times are sent in 12hr format instead of 24, making the API throw an error stating that times are not in chronological order. This is not an issue when creating a medication, however.
